Updates the edition guide to match the latest changes (#1062)
ember-learn/guides-source: guides/release/upgrading/editions.md

attributes. The component you are invoking decides where to put these attributes
by using the special `...attributes` syntax. This will be discussed later in
the section on components.

For classic components, only attributes that were explicitly listed by the component
you are invoking would be placed on the component's wrapper element.

- _Arguments_ and _attributes_ are distinguished from each other when using a
  component. With curly brace style components, every value you pass to the
  component is an _argument_ - a JS value that is passed to the component class
  so it can be used in the component's template:

  ```handlebars
  {{!-- In this example, `value` is an argument --}}

>   {{input value=this.text}}

Reviewer: Not sure if `{{input` is a good example, since it's not really a component since no dash.

Reviewer: Plus `@value` vs `value` comes up right away since it's built in.

  ```

  With angle brackets, since you can pass standard HTML attributes to the
  component directly, we need a way to distinguish between those and the
  component's arguments. To do this, we use the `@` symbol:

  ```handlebars

>   <CustomInput @value={{this.text}}>

Reviewer: ```suggestion
  <CustomInput @value={{this.text}}/>
```

  ```

  This allows you to see at a glance whether a value is an argument, which will
  likely affect the JS of a component, or an attribute, which will likely affect
  the HTML of a component.

#### Getting used to Angle Brackets

Here are the main differences between angle bracket and curly syntax:

- The component name is in `CapitalCase` instead of `kebab-case`.
  `{{my-component}}` becomes `<MyComponent />`.

- Components open and close in the same way as HTML elements. Components that do
  not accept a block can use the self closing syntax (a trailing slash) just
  like `<img />` or other tags.

- _Arguments_ are passed by adding `@` to the front of the argument name:

  ```handlebars
    {{!-- Before --}}

>     {{todo item=item}}

Reviewer: ```suggestion
    {{todo-item item=item}}
```

    {{!-- After --}}

>     <Todo @item={{item}}/>

Reviewer: ```suggestion
    <Todo @item={{this.item}}/>
```

Author: Great catches, thanks for reviewing 😄

  ```

- When you pass a bound value to a component, remember that it needs to be
  wrapped in curly braces:

  ```handlebars

>     <Todo @done={{isDone}}/>

Reviewer: ```suggestion
    <Todo @done={{this.isDone}}/>
```

}
```

The `init` hook still exists on many existing classes, and runs _after_
`constructor`, so you can generally convert to native class syntax without
rewriting your `init` methods. However, in the future `init` will be removed,
so you should eventually transition to `constructor`.

> 

Reviewer: Should `@classic` be mentioned?

Author: Possibly, I think it'd be best to add in a followup though

This will push the tag for the `[]` property onto the autotrack stack, and that
property is what is invalidated when the array is updated with KVO methods.

> ## Glimmer Componentts

Reviewer 2: ```suggestion
## Glimmer Components
```

`@ember`. Behind the scenes, Ember's rendering engine comes from Glimmer, and
now the components do too. However, Glimmer is a low-level integration with
Ember, and you don't need to worry about learning it separately.

### Benefits of Glimmer Components

Glimmer components have a some huge benefits

- These new components give you all the benefits described in Native Classes
  above
- They don't extend from `EmberObject` at all, which means that they don't need
  `EmberObject` APIs, such as `reopenClass`, `extend`. You can safely use
  `constructor` for all setup code.
- Lifecycle hooks are greatly simplified and easier to use
- They don't have that wrapping HTML element that got in the way of CSS styling
  and layout

Arguments are also namespaced on `this.args` within Glimmer components, which is
an immutable object. This means that:

- It's clear when you are accessing arguments passed to the component, and
  when you are accessing fields and properties of the component itself.
- Arguments always refer to the original value that was passed in, so you
  don't have to track down confusing code in hooks or computed property
  definitions that modifies the value of the argument.

> - There is no confusing 2-way-databinding for arguments via the component

Reviewer 2: ```suggestion
- There is no confusing two-way data binding for arguments via the component
```

To sum it up, the new mental model is that the "wrapping" element is just like
any other element in your template, and you interact with it in exactly the same
way. This means that when converting a classic component, you will need to add
the wrapping element that was there previously to the template (unless it was a
tagless component, e.g. `tagName: ''`).

#### Mixins

These components do _not_ support Ember mixins. Before native classes were
available in JavaScript, mixins gave Ember developers some powers that are
similar to class inheritance. For apps that use mixins, the recommended path is

> to refactor the ixins to be native classes instead, which the other parts of

Reviewer: ```suggestion
to refactor the mixins to be native classes instead, which the other parts of
```

> your app can inherit from.

Reviewer: Also worth mentioning using container/provider component patterns.

  @tracked scrollOffset;

  @action
  listener(e) {
    this.scrollOffset = e.clientY;
  }
}
```

```handlebars {data-filename=app/components/scroll-component.hbs}
<div {{add-event-listener "scroll" this.listener}}>
  ...
</div>
```

This modifier generalizes the functionality that the component implemented using
lifecycle hooks before, so we can use this modifier whenever we need to in _any_
component. This is a much better solution than manually managing event listeners
every time we need one! At this point, the modifier is effectively the same as
the `{{on}}` modifier as well, so we could get rid of it altogether and replace
it with `on`:

```handlebars {data-filename=app/components/scroll-component.hbs}
<div {{on "scroll" this.listener}}>
  ...
</div>
```

### Template-Only Components

In Octane, template-only components have only an `hbs` file and no `JavaScript` file.
Behind the scenes, template-only components inherit from `'@glimmer/component'`.

They can be thought of as _functional_ components, in the sense that their
output (the rendered template) is a pure function of their inputs (their
arguments). The fact that they can't have state makes them much easier to reason

> about in general, and less prone to errors.

Reviewer: probably worth mentioning that `@args` are what you would use
